### PR TITLE
Feat/database

### PR DESCRIPTION
**Pull request by Andriy-Shymkiv**
## Summary by CodeRabbit

- **New Features**
  - Introduced a search functionality to the app.

- **Documentation**
  - Added styles for the search bar.

- **Refactor**
  - Added a `Search` class with methods for fetching and displaying results.
  - Added a search bar at the top of the `Hero` component and a `Search` component to the `App` component.

### CODE REVIEW

File: src/database/database.service.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  async connect(): Promise<void> {
+    try {
+      await this.pool.connect();
+      this.logger.log('Connected to database');
+    } catch (error) {
+      this.logger.error('Error connecting to database:', error);
+    }
+  }
+
+  async disconnect(): Promise<void> {
```
Comment (KanievskyiDanylo) [inline src/database/database.service.ts:27]: Run this for `onApplicationShutdown`. Check this [doc](https://docs.nestjs.com/fundamentals/lifecycle-events#application-shutdown)

File: src/database/database.service.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import { Pool } from 'pg';
+
+@Injectable()
+export class DatabaseService implements OnApplicationBootstrap {
+  private pool: Pool;
+  private readonly logger = new Logger(DatabaseService.name);
+
+  constructor() {
+    this.pool = new Pool({
+      user: process.env.DB_USER,
```
Comment (KanievskyiDanylo) [inline src/database/database.service.ts:11]: Use env validation for these variables. Zod could be a great library for this

File: src/database/database.service.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      host: process.env.DB_HOST,
+      database: process.env.DB_NAME,
+      port: process.env.DB_PORT ? parseInt(process.env.DB_PORT) : undefined,
```
Comment (KanievskyiDanylo) [inline src/database/database.service.ts:14]: It's better to do such stuff in EnvValidation, where you can set default values if user didn't provide env variables

File: src/database/database.service.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    } catch (error) {
+      this.logger.error('Error connecting to database:', error);
+    }
+  }
+
+  async disconnect(): Promise<void> {
+    await this.pool.end();
+  }
+
+  async query(query: string, params?: any[]): Promise<any> {
```
Comment (KanievskyiDanylo) [inline src/database/database.service.ts:31]: Add generic type, it could be useful for working with entities

Comment (KanievskyiDanylo) [reply src/database/database.service.ts:31]: Also specify Error in return type

File: src/database/database.service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { DatabaseService } from './database.service';
+
+describe('DatabaseService', () => {
+  let service: DatabaseService;
+
+  beforeEach(async () => {
```
Comment (KanievskyiDanylo) [inline src/database/database.service.spec.ts:7]: Add test for basic queries (create, update, select)

Comment (Andriy-Shymkiv) [reply src/database/database.service.spec.ts:7]: will do it on a separate PR

File: src/config/env.validation.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      DB_PORT: Number(config.DB_PORT),
+    });
+    return validatedConfig;
+  } catch (error: any) {
+    throw new Error(error.message);
+  }
+}
+
+// can't immediately call validate(process.env) because process.env is not yet defined
+export const env = (): ValidatedEnv => validate(process.env);
```
Comment (KanievskyiDanylo) [inline src/config/env.validation.ts:33]: No need to call it each time, just do it once

File: src/database/database.service.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  private readonly logger = new Logger(DatabaseService.name);
+
+  constructor() {
+    this.pool = new Pool({
+      user: env().DB_USER,
+      host: env().DB_HOST,
+      database: env().DB_NAME,
+      port: env().DB_PORT,
+    });
+  }
```
Comment (KanievskyiDanylo) [inline src/database/database.service.ts:26]: Check the comment above about calling `env()` each time (no need to do it each time)

File: src/database/database.service.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+      }
+    }
+  }
+
+  async onApplicationBootstrap(): Promise<void> {
+    await this.connect();
+  }
+
+  async onApplicationShutdown(): Promise<void> {
+    await this.disconnect();
```
Comment (KanievskyiDanylo) [inline src/database/database.service.ts:75]: I am not sure that this would be enough, have you been able to test this hook?

Comment (KanievskyiDanylo) [reply src/database/database.service.ts:75]: 1. https://stackoverflow.com/questions/61892684/nest-js-gracefully-shutdown
2. 
image

File: .gitignore [hunk 9]
```diff
@@ -28,7 +28,7 @@ lerna-debug.log*
 *.sublime-workspace
 
 # IDE - VSCode
-/vscode
+/.vscode
```
Comment (KanievskyiDanylo) [inline .gitignore:31]: Seems redundant

File: src/database/database.service.ts [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        'code' in error
+      ) {
+        // postgresql error
+        throw new BadRequestException(
+          `Error code: ${(error as { code: string }).code}`,
+        );
+      } else {
+        // unexpected error
+        throw new InternalServerErrorException();
+      }
```
Comment (KanievskyiDanylo) [inline src/database/database.service.ts:66]: You can do it without `if -> else-if -> else`, just do `if -> if -> return`

File: src/config/env.validation.ts (hunk 6 above)
Comment (KanievskyiDanylo) [reply src/config/env.validation.ts:33]: About comment above, than it means that you should call validatation after process.env is loaded (I assume it should be after `GlobalConfig.forRoot()`

File: src/database/database.service.ts (hunk 8 above)
Comment (Andriy-Shymkiv) [reply src/database/database.service.ts:75]: `onApplicationShutdown` works properly, I just checked the number of active connections before and after shutdown - https://stackoverflow.com/questions/5267715/right-query-to-get-the-current-number-of-connections-in-a-postgresql-db

File: src/config/env.validation.ts (hunk 6 above)
Comment (Andriy-Shymkiv) [reply src/config/env.validation.ts:33]: I tried many times but it didn't work except to put the `validate` function call right after `ConfigModule.forRoot({})` (in the same file)

Comment (Andriy-Shymkiv) [reply src/config/env.validation.ts:33]: I tried many times but it won't work except of just putting the `validate` function call right after the `ConfigModule.forRoot({})` (same file)

**Pull request merged**